File: frontend/text_segmentation.js
// Canonical non-workflow frontend sentence policy.  It mirrors the backend
// core scanner, while preserving exact separator tokens for UI diffs.  Workflow
// plugins deliberately do not import this module.

const TERMINATORS = new Set([...".!?…。！？؟۔｡．।॥"]);
const TIGHT_TERMINATORS = new Set([..."…。！？؟۔｡．।॥"]);
const TRAILING_MARKERS = new Set([..."'»›」』*_)]}>”’“‘\""]);

const ALWAYS_NONTERMINAL = new Set(["e.g.", "i.e.", "a.k.a.", "vs.", "v.", "cf."]);
const LOWERCASE_CONTINUATION = new Set([
  "a.m.",
  "p.m.",
  "approx.",
  "dept.",
  "est.",
  "misc.",
  "incl.",
  "esp.",
  "min.",
  "max.",
  "ref.",
  "sec.",
]);
const TITLES = new Set([
  "mr.",
  "mrs.",
  "ms.",
  "mx.",
  "dr.",
  "prof.",
  "rev.",
  "hon.",
  "pres.",
  "gov.",
  "sen.",
  "rep.",
  "sr.",
  "jr.",
  "st.",
  "mt.",
  "capt.",
  "cpt.",
  "lt.",
  "col.",
  "gen.",
  "sgt.",
  "adm.",
  "maj.",
]);
const NUMBER_ABBREVIATIONS = new Set([
  "no.",
  "fig.",
  "eq.",
  "ch.",
  "vol.",
  "pp.",
  "jan.",
  "feb.",
  "mar.",
  "apr.",
  "jun.",
  "jul.",
  "aug.",
  "sep.",
  "sept.",
  "oct.",
  "nov.",
  "dec.",
]);

const LETTER = /\p{L}/u;
const ALNUM = /[\p{L}\p{N}]/u;
const ABBREVIATION = /(?:\p{L}+\.)+$/u;

export function isHardLineBreak(c) {
  return (
    c === "\n" ||
    c === "\v" ||
    c === "\f" ||
    c === "\r" ||
    c === "\x85" ||
    c === "\u2028" ||
    c === "\u2029" ||
    (c >= "\x1c" && c <= "\x1e")
  );
}

export function isSentenceWhitespace(c) {
  return c != null && /\s/u.test(c);
}

function _isUpper(c) {
  return Boolean(c && LETTER.test(c) && c === c.toUpperCase() && c !== c.toLowerCase());
}

function _periodIsNonterminal(text, period, nextIndex) {
  if (/\d/u.test(text[period - 1] || "") && /\d/u.test(text[period + 1] || "")) return true;
  const match = text.slice(0, period + 1).match(ABBREVIATION);
  if (!match) return false;
  const raw = match[0];
  const abbreviation = raw.toLowerCase();
  const next = text[nextIndex] || "";
  if (ALWAYS_NONTERMINAL.has(abbreviation)) return true;
  if (TITLES.has(abbreviation) && ALNUM.test(next)) return true;
  if (NUMBER_ABBREVIATIONS.has(abbreviation) && /\d/u.test(next)) return true;
  const letters = raw.replaceAll(".", "");
  if (
    ALNUM.test(next) &&
    (letters.length === 1 || ((abbreviation.match(/\./g) || []).length >= 2 && letters === letters.toUpperCase()))
  ) {
    return true;
  }
  return LOWERCASE_CONTINUATION.has(abbreviation) && next === next.toLowerCase() && LETTER.test(next);
}

export function sentenceBoundaryEnds(text) {
  const ends = [];
  let i = 0;
  while (i < text.length) {
    if (!TERMINATORS.has(text[i])) {
      i += 1;
      continue;
    }
    let terminalEnd = i + 1;
    while (terminalEnd < text.length && TERMINATORS.has(text[terminalEnd])) terminalEnd += 1;
    let markerEnd = terminalEnd;
    while (markerEnd < text.length && TRAILING_MARKERS.has(text[markerEnd])) markerEnd += 1;
    let boundaryEnd = markerEnd;
    while (
      boundaryEnd < text.length &&
      isSentenceWhitespace(text[boundaryEnd]) &&
      !isHardLineBreak(text[boundaryEnd])
    ) {
      boundaryEnd += 1;
    }
    const hasSeparator = boundaryEnd > markerEnd;
    const terminals = text.slice(i, terminalEnd);
    const allowsTight =
      [...terminals].some((c) => TIGHT_TERMINATORS.has(c)) ||
      (markerEnd < text.length && _isUpper(text[markerEnd]) && [...terminals].some((c) => ".!?".includes(c)));
    if (!hasSeparator && !(allowsTight && markerEnd < text.length)) {
      i = terminalEnd;
      continue;
    }
    const onePeriod = terminalEnd === i + 1 && text[i] === ".";
    if (!(onePeriod && _periodIsNonterminal(text, i, boundaryEnd))) {
      ends.push(boundaryEnd);
      i = boundaryEnd;
    } else {
      i = terminalEnd;
    }
  }
  return ends;
}

function _lineStream(line) {
  const out = [];
  let start = 0;
  for (const end of sentenceBoundaryEnds(line)) {
    if (line.slice(start, end).trim()) out.push({ kind: "sentence", text: line.slice(start, end) });
    else if (end > start) out.push({ kind: "spacing", text: line.slice(start, end) });
    start = end;
  }
  if (start < line.length) {
    out.push({ kind: line.slice(start).trim() ? "sentence" : "spacing", text: line.slice(start) });
  }
  return out;
}

// Lossless token stream: sentence entries never contain a line break; every
// line separator is its own entry. Joining all entry text reproduces input.
export function sentenceStream(text) {
  const out = [];
  let start = 0;
  for (let i = 0; i < text.length; i++) {
    if (!isHardLineBreak(text[i])) continue;
    out.push(..._lineStream(text.slice(start, i)));
    let end = i + 1;
    if (text[i] === "\r" && text[end] === "\n") end += 1;
    out.push({ kind: "linebreak", text: text.slice(i, end) });
    i = end - 1;
    start = end;
  }
  out.push(..._lineStream(text.slice(start)));
  return out;
}

export function splitSentences(text) {
  return sentenceStream(text)
    .filter((unit) => unit.kind === "sentence")
    .map((unit) => unit.text.trim())
    .filter(Boolean);
}

export function endsWithSentenceTerminator(text) {
  let trimmed = text.trimEnd();
  while (trimmed && TRAILING_MARKERS.has(trimmed.at(-1))) trimmed = trimmed.slice(0, -1).trimEnd();
  return Boolean(trimmed && TERMINATORS.has(trimmed.at(-1)));
}

// DOM segmentation sees whitespace-separated visual words. Probe the same
// scanner with the next word so abbreviation decisions stay identical.
export function tokenEndsSentence(token, nextToken = "") {
  if (!token || !nextToken) return false;
  const probe = `${token} ${nextToken}`;
  const first = sentenceBoundaryEnds(probe)[0];
  return first != null && first <= token.length + 1;
}

export function splitTightSentenceChunks(token) {
  const out = [];
  let start = 0;
  for (const end of sentenceBoundaryEnds(token)) {
    if (end > start) out.push(token.slice(start, end));
    start = end;
  }
  if (start < token.length) out.push(token.slice(start));
  return out.length ? out : [token];
}

File: backend/core/text_segmentation.py
"""Dependency-free sentence and quotation primitives."""

from __future__ import annotations

import re
from collections.abc import Iterator

__all__ = [
    "CLOSE_QUOTES",
    "HARD_LINE_BREAK_RE",
    "OPEN_QUOTES",
    "PARA_SPLIT",
    "SENT_SPLIT",
    "TOGGLE_QUOTES",
    "count_sentences",
    "ends_with_question",
    "ends_with_sentence_terminator",
    "extract_unquoted_text",
    "find_quote_spans",
    "remove_quoted_spans",
    "sentence_boundary_ends",
    "split_paragraphs",
    "split_sentence_units",
    "split_sentences",
]

# Python ``str.splitlines`` recognizes this complete set.  Keep an explicit
# pattern for callers that need to preserve or inspect the separators.
HARD_LINE_BREAK_RE = re.compile(r"\r\n|[\n\v\f\r\x1c-\x1e\x85\u2028\u2029]")
PARA_SPLIT = re.compile(r"(?:\r\n|[\n\r\x85\u2028\u2029])\s*(?:\r\n|[\n\r\x85\u2028\u2029])")

# Compatibility pattern for older consumers that call ``SENT_SPLIT.split``.
# Public splitting uses the scanner below for abbreviations and Unicode marks.
SENT_SPLIT = re.compile(
    r"(?:(?<=[.!?…])[\"\u201d\u2019'*_)\]]*\s+|(?:\r\n|[\n\v\f\r\x1c-\x1e\x85\u2028\u2029])+)",
)

_QUOTE_PAIRS = {
    "“": "”",
    "‘": "’",
    "«": "»",
    "‹": "›",
    "「": "」",
    "『": "』",
    "„": "“",
    "‚": "‘",
}
OPEN_QUOTES = frozenset(_QUOTE_PAIRS)
CLOSE_QUOTES = frozenset(_QUOTE_PAIRS.values())
TOGGLE_QUOTES = frozenset({'"'})

_TERMINATORS = frozenset(".!?…。！？؟۔｡．।॥")
_QUESTION_TERMINATORS = frozenset("?？؟")
_TIGHT_TERMINATORS = frozenset("…。！？؟۔｡．।॥")
_TRAILING_MARKERS = frozenset("'»›」』*_)]}>") | CLOSE_QUOTES | TOGGLE_QUOTES
_TRAILING_TERMINALS = "!.…。！۔｡．।॥"

_ALWAYS_NONTERMINAL_ABBREVIATIONS = frozenset({"e.g.", "i.e.", "a.k.a.", "vs.", "v.", "cf."})
_LOWERCASE_CONTINUATION_ABBREVIATIONS = frozenset(
    {"a.m.", "p.m.", "approx.", "dept.", "est.", "misc.", "incl.", "esp.", "min.", "max.", "ref.", "sec."}
)
_TITLE_ABBREVIATIONS = frozenset(
    {
        "mr.",
        "mrs.",
        "ms.",
        "mx.",
        "dr.",
        "prof.",
        "rev.",
        "hon.",
        "pres.",
        "gov.",
        "sen.",
        "rep.",
        "sr.",
        "jr.",
        "st.",
        "mt.",
        "capt.",
        "cpt.",
        "lt.",
        "col.",
        "gen.",
        "sgt.",
        "adm.",
        "maj.",
    }
)
_NUMBER_ABBREVIATIONS = frozenset(
    {
        "no.",
        "fig.",
        "eq.",
        "ch.",
        "vol.",
        "pp.",
        "jan.",
        "feb.",
        "mar.",
        "apr.",
        "jun.",
        "jul.",
        "aug.",
        "sep.",
        "sept.",
        "oct.",
        "nov.",
        "dec.",
    }
)
_ABBREVIATION_BEFORE_PERIOD = re.compile(r"(?:[^\W\d_]+\.)+$", re.UNICODE)


def _period_is_nonterminal(text: str, period: int, next_char: int) -> bool:
    if period > 0 and period + 1 < len(text) and text[period - 1].isdigit() and text[period + 1].isdigit():
        return True

    match = _ABBREVIATION_BEFORE_PERIOD.search(text[: period + 1])
    if match is None:
        return False
    raw_abbreviation = match.group(0)
    abbreviation = raw_abbreviation.casefold()
    next_value = text[next_char] if next_char < len(text) else ""

    if abbreviation in _ALWAYS_NONTERMINAL_ABBREVIATIONS:
        return True
    if abbreviation in _TITLE_ABBREVIATIONS and next_value.isalnum():
        return True
    if abbreviation in _NUMBER_ABBREVIATIONS and next_value.isdigit():
        return True

    letters = raw_abbreviation.replace(".", "")
    if next_value.isalnum() and (len(letters) == 1 or (abbreviation.count(".") >= 2 and letters.isupper())):
        return True
    return abbreviation in _LOWERCASE_CONTINUATION_ABBREVIATIONS and next_value.islower()


def sentence_boundary_ends(text: str) -> Iterator[int]:
    """Yield lossless exclusive ends of punctuation-complete sentences.

    This iterator does not manufacture a sentence object and may include
    separator whitespace in its offsets.  Use :func:`split_sentence_units`
    when consuming sentence text; that function removes all line separators.
    """
    i = 0
    size = len(text)
    while i < size:
        if text[i] not in _TERMINATORS:
            i += 1
            continue

        terminal_end = i + 1
        while terminal_end < size and text[terminal_end] in _TERMINATORS:
            terminal_end += 1

        marker_end = terminal_end
        while marker_end < size and text[marker_end] in _TRAILING_MARKERS:
            marker_end += 1

        boundary_end = marker_end
        while boundary_end < size and text[boundary_end].isspace():
            boundary_end += 1
        has_separator = boundary_end > marker_end
        allows_tight_boundary = any(ch in _TIGHT_TERMINATORS for ch in text[i:terminal_end]) or (
            marker_end < size and text[marker_end].isupper() and any(ch in ".!?" for ch in text[i:terminal_end])
        )
        if not has_separator and not (allows_tight_boundary and marker_end < size):
            i = terminal_end
            continue

        only_one_period = terminal_end == i + 1 and text[i] == "."
        if not (only_one_period and _period_is_nonterminal(text, i, boundary_end)):
            yield boundary_end
            i = boundary_end
        else:
            i = terminal_end


def _split_sentence_line(text: str) -> list[str]:
    units: list[str] = []
    start = 0
    for end in sentence_boundary_ends(text):
        unit = text[start:end].strip()
        if unit:
            units.append(unit)
        start = end
    tail = text[start:].strip()
    if tail:
        units.append(tail)
    return units


def split_sentence_units(text: str) -> list[str]:
    """Split text into sentence units with line breaks as hard separators."""
    units: list[str] = []
    for line in text.splitlines():
        units.extend(_split_sentence_line(line))
    return units


def ends_with_sentence_terminator(text: str) -> bool:
    trimmed = text.rstrip()
    while trimmed and trimmed[-1] in _TRAILING_MARKERS:
        trimmed = trimmed[:-1].rstrip()
    return bool(trimmed and trimmed[-1] in _TERMINATORS)


def split_paragraphs(text: str) -> list[str]:
    """Split on blank-line paragraph breaks, dropping empty paragraphs."""
    return [paragraph for paragraph in PARA_SPLIT.split(text.strip()) if paragraph.strip()]


def split_sentences(text: str) -> list[str]:
    """Split text into sentences; no returned value contains a line break."""
    sentences: list[str] = []
    for paragraph in split_paragraphs(text):
        sentences.extend(split_sentence_units(paragraph))
    return sentences


def find_quote_spans(text: str) -> list[tuple[int, int]]:
    """Return balanced maximal dialogue spans, including their outer marks."""
    spans: list[tuple[int, int]] = []
    stack: list[str] = []
    outer_start = 0
    for i, ch in enumerate(text):
        if ch in TOGGLE_QUOTES:
            slashes = 0
            j = i - 1
            while j >= 0 and text[j] == "\\":
                slashes += 1
                j -= 1
            if slashes % 2:
                continue

        if ch == "’" and i > 0 and i + 1 < len(text) and text[i - 1].isalnum() and text[i + 1].isalnum():
            continue
        if ch in TOGGLE_QUOTES and not stack and i > 0 and text[i - 1].isdigit():
            continue

        if stack and ch == stack[-1]:
            stack.pop()
            if not stack:
                spans.append((outer_start, i + 1))
            continue

        if ch in TOGGLE_QUOTES:
            if not stack:
                outer_start = i
            stack.append(ch)
        elif ch in OPEN_QUOTES:
            if not stack:
                outer_start = i
            stack.append(_QUOTE_PAIRS[ch])
        elif ch in CLOSE_QUOTES and stack:
            if ch in stack:
                while stack:
                    if stack.pop() == ch:
                        break
            if not stack:
                spans.append((outer_start, i + 1))
    return spans


def extract_unquoted_text(text: str) -> str:
    """Remove balanced quoted spans and normalize surrounding whitespace."""
    return " ".join(remove_quoted_spans(text).split())


def remove_quoted_spans(text: str) -> str:
    """Remove balanced quoted spans while preserving all outside characters."""
    pieces: list[str] = []
    cursor = 0
    for start, end in find_quote_spans(text):
        pieces.append(text[cursor:start])
        pieces.append(" ")
        cursor = end
    pieces.append(text[cursor:])
    return "".join(pieces)


def count_sentences(text: str) -> int:
    return len(split_sentence_units(text.strip())) if text.strip() else 0


def ends_with_question(sentence: str) -> bool:
    trimmed = sentence.rstrip().rstrip("".join(_TRAILING_MARKERS)).rstrip(_TRAILING_TERMINALS)
    return bool(trimmed and trimmed[-1] in _QUESTION_TERMINATORS)
